crest/Assets/Crest/Crest/Shaders/Underwater/UnderwaterMaskShared.hlsl

```diff
@@ -101,6 +101,16 @@ Varyings Vert(Attributes v)
 
 half4 Frag(const Varyings input, const bool i_isFrontFace : SV_IsFrontFace) : SV_Target
 {
+#if CREST_HANDLE_XR
+	// @MSAAOutlineFix:
+	// The edge of the ocean surface at the near plane will be MSAA'd leaving a noticeable edge. By rendering the mask
+	// with a slightly further near plane, it exposes the edge to having the underwater fog applied which is much nicer.
+	if (_CrestDepthTextureOffset > 0 && LinearEyeDepth(input.positionCS.z) < (_ProjectionParams.y + 0.001))
+	{
+		discard;
+	}
+#endif
```

Comment:
@huwb For XR SPI I am doing it in the shader as XR SPI doesn't accept changing the matrices.

Reply:
ok, is it worth adding this note as a comment?

Reply:
Can do. Actually... I probably should only do it this way. I haven't noticed any issues, but I just realised that changing the near plane in C# will mess with LinearEyeDepth working correctly since it uses _ZBufferParams which should be same as the main depth buffer. Thoughts?

Reply:
> i feel that _ZBufferParams should be correct if you change the near plane?

It's only for this pass which is rendered into its own depth texture. I was thinking when sampled later on the _ZBufferParams values will be computed from the non modified near far plane.

Another issue illustrated in the following screenshots is the underwater effect rendering the device depth to the screen. One with no near plane offset and one with a 10 near plane offset. The one with the modified near plane produces a depth value which is incorrect in context. It's an extreme case so it might be fine with the very small offset (0.001) I am using.

![111](https://user-images.githubusercontent.com/5249806/131210984-07b0d56f-fa92-4f30-a3fe-890033442d45.jpg)
Ocean mask depth texture rendered (0 near plane offset)

![122](https://user-images.githubusercontent.com/5249806/131210987-df8cd74d-55c5-429b-8632-2f5dc33220bf.jpg)
Ocean mask depth texture rendered (10 near plane offset)

Hopefully it makes sense.

Reply:
Went with doing it in the shader only and removing the C# one due to above issues.